c-bindings-gen/src/main.rs

```diff
@@ -0,0 +1,809 @@
+
+	let mut out = std::fs::OpenOptions::new().write(true).create(true).truncate(true)
+		.open(&out_path).expect("Unable to open new src file");
+
+	assert_eq!(export_status(&syntax.attrs), ExportStatus::Export);
+	println_docs(&mut out, &syntax.attrs, "");
+
+	let mut type_resolver = TypeResolver::new(module);
+	let orig_module = orig_crate.to_string() + "::" + module;
+
+	if path.ends_with("lib.rs") {
+		write!(out, "#![allow(non_camel_case_types)]\n").unwrap();
+		write!(out, "#![allow(non_snake_case)]\n").unwrap();
+		write!(out, "#![allow(unused_imports)]\n").unwrap();
+		write!(out, "#![allow(unused_variables)]\n").unwrap();
+		write!(out, "mod c_types;\n").unwrap();
+		write!(out, "mod bitcoin;\n").unwrap();
+	} else {
+		write!(out, "\nuse std::ffi::c_void;\nuse bitcoin::hashes::Hash;\n\n").unwrap();
+	}
+
+	for item in syntax.items.iter() {
+		match item {
+			syn::Item::Use(u) => type_resolver.process_use(&mut out, &u),
+			syn::Item::Static(_) => {},
+			syn::Item::Enum(e) => {
+				if let syn::Visibility::Public(_) = e.vis {
+					println_enum(&mut out, &e, &orig_module, &mut type_resolver, header_file);
+				}
+			},
+			syn::Item::Impl(i) => {
+				println_impl(&mut out, &i, &type_resolver);
+			},
+			syn::Item::Struct(s) => {
+				if let syn::Visibility::Public(_) = s.vis {
+					println_struct(&mut out, &s, &orig_module, &mut type_resolver, header_file);
+				}
+			},
+			syn::Item::Trait(t) => {
+				if let syn::Visibility::Public(_) = t.vis {
+					println_trait(&mut out, &t, &orig_module, &mut type_resolver);
+				}
+			},
+			syn::Item::Mod(m) => {
+				if let syn::Visibility::Public(_) = m.vis {
+					match export_status(&m.attrs) {
+						ExportStatus::Export => {},
+						ExportStatus::NoExport|ExportStatus::TestOnly => continue,
+						ExportStatus::Rename(_) => unimplemented!(),
+					}
+
+					if m.content.is_some() { unimplemented!(); } // Probably mod tests {}
+					let f_path = format!("{}/{}.rs", (path.as_ref() as &Path).parent().unwrap().display(), m.ident);
+					let new_mod = if module.is_empty() { format!("{}", m.ident) } else { format!("{}::{}", module, m.ident) };
+					if let Ok(_) = File::open(&f_path) {
+
+						if should_export(&f_path) {
+							println_docs(&mut out, &m.attrs, "");
+							write!(out, "pub mod {};\n", m.ident).unwrap();
+							convert_file(&f_path, &format!("{}/{}.rs", (out_path.as_ref() as &Path).parent().unwrap().display(), m.ident),
+								orig_crate, &new_mod, header_file);
+						}
+					} else {
+						let f_path = format!("{}/{}/mod.rs", (path.as_ref() as &Path).parent().unwrap().display(), m.ident);
+						if should_export(&f_path) {
+							println_docs(&mut out, &m.attrs, "");
+							write!(out, "pub mod {};\n", m.ident).unwrap();
+							convert_file(&f_path, &format!("{}/{}/mod.rs", (out_path.as_ref() as &Path).parent().unwrap().display(), m.ident),
+								orig_crate, &new_mod, header_file);
+						}
+					}
```

Comment:
Let's remove the duplication by extracting a helper for opening the file. Or better yet move this logic to the beginning of `convert_file`.